sql/src/main/java/io/crate/analyze/symbol/SymbolFormatter.java

```diff
@@ -54,65 +107,269 @@ public static String format(String messageTmpl, Symbol ... symbols) {
             if (s == null) {
                 formattedSymbols[i] = "null";
             } else {
-                formattedSymbols[i] = s.accept(INSTANCE, null);
+                formattedSymbols[i] = formatSimple(s);
             }
         }
         return String.format(Locale.ENGLISH, messageTmpl, formattedSymbols);
     }
 
-    private static String format(String tmpl, Object ... args) {
-        return String.format(Locale.ENGLISH, tmpl, args);
+
+
+    public String format(Symbol symbol) {
+        return format(symbol, Context.DEFAULT_MAX_DEPTH, true);
+    }
+
+    /**
+     * format the given symbol
+     * @param symbol the symbol to format
+     * @param maxDepth the max depth to print, if maxDepth is reached, "..." is printed for the rest
+     * @param fullQualified if references should be fully qualified (contain schema and table name)
+     * @return
+     */
+    public String format(Symbol symbol, int maxDepth, boolean fullQualified) {
+        Context context = new Context(maxDepth, fullQualified);
+        symbol.accept(this, context);
+        return context.formatted();
+    }
+
+    /**
+     * format without special treatment of operator or other special function
+     */
+    public static String formatSimple(Symbol symbol) {
+        return formatSimple(symbol, Context.DEFAULT_MAX_DEPTH, false);
+    }
+
+    public static String formatFullQualified(Symbol symbol) {
+        return formatSimple(symbol, Context.DEFAULT_MAX_DEPTH, true);
     }
 
-    public static String format(Symbol symbol) {
-        return symbol.accept(INSTANCE, null);
+    public static String formatSimple(Symbol symbol, int maxDepth, boolean fullQualified) {
+        return INSTANCE.format(symbol, maxDepth, fullQualified);
     }
 
     @Override
-    protected String visitSymbol(Symbol symbol, Void context) {
-        return symbol.toString();
+    protected Void visitSymbol(Symbol symbol, Context context) {
+        if (context.maxDepthReached()) {
+            context.addEllipsis();
```

Comment:
would prefer adding ellipsis through a given flag, otherwise an exception should be thrown if maxDepth is reached. This will prevent adding an invalid expression to the metaData which could lead then to a cluster-fuck-up on reading metaData